File: parser.fs
type SExp =
    | Atom of string
    | SList of SExp list

let rec stringOfSExp sexp =
    match sexp with
    | Atom v -> v
    | SList l -> "(" + String.concat "" (List.map stringOfSExp l) + ")"

let tokenize input =
    input.Replace("(", " ( ").Replace(")", " ) ").Split(" ", System.StringSplitOptions.RemoveEmptyEntries) |> List.ofArray

let rec parseTokens tokens =
    match tokens with
    | [] -> Error "Unexpected end of input"
    | token :: rest ->
        match token with
        | "(" ->
            let result = parseList rest
            match result with
            | Ok (children, rest) -> Ok (SList children, rest)
            | Error e -> Error e
        | ")" -> Error "Unmatched )"
        | _ -> Ok (Atom token, rest)
and parseList tokens =
    match tokens with
    | [] -> Error "Unmatched ("
    | ")" :: rest -> Ok ([], rest)
    | _ ->
        match parseTokens tokens with
        | Ok (sexp, rest) ->
            match parseList rest with
            | Ok (sexps, finalRest) -> Ok (sexp :: sexps, finalRest)
            | Error e -> Error e
        | Error e -> Error e

let parse input =
    let tokens = tokenize input
    match parseTokens tokens with
    | Ok (sexp, []) -> sexp
    | Ok _ -> failwith "Invalid S-expression"
    | Error e -> failwith e

[<EntryPoint>]
let main _ =
    let input = "(+ 1 (* 2 3))"
    let result = parse input
    printfn "%s" (stringOfSExp result)
    0

File: parser.ml
type sexp =
  | Atom of string
  | SList of sexp list

let rec string_of_sexp sexp =
  match sexp with
  | Atom v -> v
  | SList l -> "(" ^ String.concat "" (List.map string_of_sexp l) ^ ")"

let tokenize input =
  let input = Str.global_replace (Str.regexp "(") " ( " input in
  let input = Str.global_replace (Str.regexp ")") " ) " input in
  List.filter (fun s -> String.length s > 0) (String.split_on_char ' ' input)

let rec parse_tokens tokens =
  match tokens with
  | [] -> raise (Failure "Unexpected end of input")
  | token :: rest ->
    match token with
    | "(" ->
      let children, rest = parse_list rest in
      SList children, rest
    | ")" -> raise (Failure "Unmatched )")
    | _ -> Atom token, rest

and parse_list tokens =
  match tokens with
  | [] -> raise (Failure "Unmatched (")
  | ")" :: rest -> [], rest
  | _ ->
    let sexp, rest = parse_tokens tokens in
    let sexps, final_rest = parse_list rest in
    sexp :: sexps, final_rest

let parse input =
  let tokens = tokenize input in
  match parse_tokens tokens with
  | sexp, [] -> sexp
  | _ -> raise (Failure "Invalid S-expression")

let () =
  let input = "(+ 1 (* 2 3))" in
  let result = parse input in
  print_endline (string_of_sexp result)
